### PR TITLE
Table of Contents

### LINKED ISSUE #85: Add TOC to documents
Our markdown parser has an option to create a Table of Contents based on the headings of a document. We should enable this option and add `toc` as a key to the API response for documents.

### PR DESCRIPTION
**Pull request by christophlocher**
Fixes #85

### CODE REVIEW

File: frontend/src/TableOfContents/index.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+function TOCEntry(props) {
+  if (props.cutoffLevel && props.level >= props.cutoffLevel) {
+    return <></>;
+  }
+
+  let slug = slugify(props.title);
+
+  let children = [];
+  if (props.children) {
+    children = props.children.map(c => <TOCEntry {...c} docTitle={props.docTitle} onClick={props.onClick} cutoffLevel={props.cutoffLevel} />);
```
Comment (mariuswilms) [inline frontend/src/TableOfContents/index.js]: Seems quite a lengthy line, maybe run this through prettier?

File: internal/ddt/node_doc.go [hunk 2]
```diff
@@ -122,6 +123,88 @@ func (d NodeDoc) Components() ([]*NodeDocComponent, error) {
...
+	r := bytes.NewReader(document)
+
+	doc, _ := html.Parse(r)
+	if err != nil {
+		return toc, err
+	}
+
+	var findHeadlineInDom func(*html.Node)
+	findHeadlineInDom = func(n *html.Node) {
+		if n.Type == html.ElementNode && (n.Data == "h1" || n.Data == "h2" || n.Data == "h3" || n.Data == "h4" || n.Data == "h5") {
```
Comment (mariuswilms) [inline internal/ddt/node_doc.go:159]: The matching code can be generalized using a regexp. What do you think is the clarity worth the performance tradeoff (see below)? With the regexp you could use a capturing group for the numeric part, saving the TrimPrefix below.


```
~/Code/_tmp > cat main_test.go 
package main

import (
	"regexp"
	"testing"
)

func TestHelloWorld(t *testing.T) {
	// t.Fatal("not implemented")
}

func BenchmarkRegexp(b *testing.B) {
	r := regexp.MustCompile(`^H[0-9]+$`)
	a := "h1"
	x := "h5"

	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		r.MatchString(a)
		r.MatchString(x)
	}
}
func BenchmarkConditional(b *testing.B) {
	a := "h1"
	x := "h5"

	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		if a == "h1" || a == "h2" || a == "h3" || a == "h4" || a == "h5" {

		}
		if x == "h1" || x == "h2" || x == "h3" || x == "h4" || x == "h5" {

		}
	}
}

go test -bench .
goos: darwin
goarch: amd64
BenchmarkRegexp-4        	10336260	       109 ns/op
BenchmarkConditional-4   	150688596	         7.95 ns/op
PASS

```

File: frontend/src/TableOfContents/TableOfContents.css [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  margin-left: 2.8rem;
+}
+
+.table-of-contents a {
+  text-decoration: none;
+  transition: all 0.2s;
+}
+
+.table-of-contents a:hover {
+  /* text-decoration: underline; */
```
Comment (mariuswilms) [inline frontend/src/TableOfContents/TableOfContents.css]:
```suggestion
```

File: internal/ddt/node_doc.go [hunk 4]
```diff
@@ -122,6 +123,88 @@ func (d NodeDoc) Components() ([]*NodeDocComponent, error) {
 	return components, nil
 }
 
+// A headline used in the ToC
```
Comment (mariuswilms) [inline internal/ddt/node_doc.go]: How do you usually handle abreviations? I try to introduce them once, or try to stick to the long form if it doesn't get too annoying. Abbreviations in code is something different.

**Review by mariuswilms — CHANGES_REQUESTED** [review]
Interested to hear your opinions. There weren't any major remarks. Thanks for adding the outlining feature!

File: frontend/src/TableOfContents/index.js (hunk 1 above)
Comment (christophlocher) [reply frontend/src/TableOfContents/index.js]: Prettier seems to keep it in a single line

File: internal/ddt/node_doc.go (hunk 2 above)
Comment (christophlocher) [reply internal/ddt/node_doc.go:159]: Personally I find the string comparison easier to read. As it is also faster and we won’t need any flexibility in the future I’d go with that.

File: internal/ddt/node_doc.go (hunk 4 above)
Comment (christophlocher) [reply internal/ddt/node_doc.go]: Good point, fixed it

**Pull request merged**